### PR TITLE
Version 1.0.1

### PR DESCRIPTION
**Pull request by lukasbestle**
## Fixes

- Staticache now handles cache values that are set by other plugins (e.g. the [Meta](https://github.com/fabianmichael/kirby-meta) plugin by @fabianmichael) more gracefully. #20

## Documentation

- Add more information to the `README` on required changes to the server setup for custom roots
- The PHP loader example code now also tries to determine the content type from the static files
- Clean up indentation in code examples

---

After merging:

- [ ] Create release
- [ ] Update https://github.com/getkirby/getkirby.com/blob/main/content/plugins/getkirby/staticache/plugin.txt with new `README` changes

### CODE REVIEW

**Comment by lukasbestle** [comment]
@jaro-io @jonasfeige @bnomei @fabianmichael Could you please test the `develop` branch to see if this fixes the issue for you?

**Comment by bnomei** [comment]
i guess it would work if the plugins do honor the array format the pages cache expects for the data, not a plain string.

CleanShot 2023-03-11 at 12 47 50

**Comment by bnomei** [comment]
CleanShot 2023-03-11 at 12 49 14

**Comment by lukasbestle** [comment]
@bnomei Thanks for testing. You are right, this requires additional changes to the plugin. See my comment in https://github.com/getkirby/staticache/issues/20#issuecomment-1407712742.

**Comment by bnomei** [comment]
apart from that it works. thanks.

File: src/StatiCache.php [hunk 1]
```diff
@@ -104,7 +105,7 @@ protected function file(string|array $key): string
 		}
 
 		$page = $kirby->page($key['id']);
-		$url  = $page->url($key['language']);
+		$url  = $page?->url($key['language']) ?? Url::to($key['id']);
 
 		// content representation paths of the home page contain the home slug
 		if ($page->isHomePage() === true && $key['contentType'] !== 'html') {
```
Comment (fabianmichael) [inline src/StatiCache.php:111]: @lukasbestle When `$page` is `null`, this still throws an error. If should be `if ($page`->isHomePage() === true […]`.

**Review by fabianmichael — CHANGES_REQUESTED** [review]

**Comment by fabianmichael** [comment]
@lukasbestle Another issue that still exists is, that the PHP-loader currently does not work for content representations, as it does not send any `content-type` header on its own … it should be possible to use e.g. PHP’s built-in [mimetype functions](https://www.php.net/manual/en/function.mime-content-type.php) to bypass the loading of Kirby’s toolkit, though I don’t have any experience with using them. Please be a ware, that this function can throw an E_WARNING that has to be dealt with:

```php
    if ($mime = @mime_content_type($path)) {
      header("Content-type: $mime");
    };
  
    die(file_get_contents($path));
```

**Comment by lukasbestle** [comment]
@fabianmichael Which PHP loader are you referring to? Staticache does not provide one, do you mean @bnomei's?

**Comment by fabianmichael** [comment]
> @fabianmichael Which PHP loader are you referring to? Staticache does not provide one, do you mean @bnomei's?

I’m talking about that loader mentioned in the README: https://github.com/getkirby/staticache/#:~:text=index.php%3F%24query_string%3B%0A%7D-,PHP%20loader%3A,-If%20you%20don%27t

File: src/StatiCache.php (hunk 1 above)
Comment (lukasbestle) [reply src/StatiCache.php:111]: Ah, you are right, I missed that line. Will fix it before the release.

**Comment by lukasbestle** [comment]
Ah, yes. Completely forgot about that one. 🙈😅

I'll see what I can do. The header mode is certainly more reliable in this case, but automatic detection could already be good enough in 99% of cases.

**Review by fabianmichael — CHANGES_REQUESTED** [review]
@lukasbestle It works much better now, but when enabling the `headers` option, it still fails with virtual pages:

```
Undefined array key "response" (StatiCache.php L58)
```

It would probably make sense use automatic testing at this point and provide a few common use cases (built-in pages, content representations, virtual pages) as test scenarios.

I am also wondering, if there are any situations, where one does not want fo use the `headers` option when using the PHP loader. I guess it’s more performant than using PHP’s built-in mime config and very likely way more reliable (I don’t have much experience with PHP’s mime functions, but there’s probably a reason why almost every CMS has its own implementation).

**Comment by lukasbestle** [comment]
TBH I feel like we are mixing different issues into the discussion. If this version solves the compatibility issues with your Meta plugin, I suggest we already publish it. We can then work on fixes for any other issues separately for 1.0.2 and in separate GitHub issues.

> It works much better now, but when enabling the headers option, it still fails with virtual pages:

Could you please open an issue for that with steps to reproduce?

> It would probably make sense use automatic testing at this point and provide a few common use cases (built-in pages, content representations, virtual pages) as test scenarios.

I agree. So far we did not find the time to write tests, but we need them anyway before we can include Staticache in the Kirby core at some point.

> I am also wondering, if there are any situations, where one does not want fo use the `headers` option when using the PHP loader. I guess it’s more performant than using PHP’s built-in mime config and very likely way more reliable (I don’t have much experience with PHP’s mime functions, but there’s probably a reason why almost every CMS has its own implementation).

We already state the performance disadvantages of the PHP loader, but I think you are right: If one wants to use the PHP loader, it is better in most cases to go for the header variant. I've added a note to the `README`.

**Comment by fabianmichael** [comment]
> Could you please open an issue for that with steps to reproduce?

@lukasbestle Thanks for the update, unfortunately I don’t really have time at the moment for dealing with additional contributions. Test setup for the issue is: Install plainkit, staticache and meta plugin, enable static cache with headers, use PHP loader and point for browser to `plainkit.whatever/sitemap.xml` to reproduce the issue.

**Comment by lukasbestle** [comment]
@fabianmichael Ah, I see. If I understand it correctly, this is not actually about virtual pages but about manually storing cache values in the page cache. See https://github.com/getkirby/staticache/pull/22#issuecomment-1464918769.

**Comment by bnomei** [comment]
its a bit about virtual pages since the sitemap would be treated like a page and be available for apache just like any other statically cached page as html document.
but yes technically its just about storing the data in cache properly.

**Comment by fabianmichael** [comment]
@lukasbestle @bnomei Would it be safer to just use a custom page model for the Sitemap and a template? That way, caching and rendering would be handled by `$page->render()` like for any other page? That would probably make the whole implementation more robust and future-proof than trying to replicate that by myself. What would you recommend? It’s probably a good idea to come up with a general solution for virtual pages and add a paragraph to the readme about how to deal with them.

**Comment by lukasbestle** [comment]
I wouldn't expect changes to Kirby's storage format for the page cache, however I cannot guarantee that we will never need to make any changes there. But if we did, it would be clearly communicated as a breaking change.

So I think both ways are fine: Either you adapt your plugin to the existing storage format for the cache or you perform a larger rewrite (which may be a breaking change for the users of your plugin though).

**Comment by bnomei** [comment]
i just tested dev branches of both plugins together and it works fine.

**Pull request merged**